File: backend/core/engine.py
# backend/core/engine.py

from backend.core.lexer import tokenize
from backend.core.parser import Parser
from backend.core.solver import solve
from backend.core.constraints import apply_constraints


def solve_equation(expression: str, constraints=None):
    # 1. Lexer
    tokens = tokenize(expression)
    print("TOKENS:", tokens)  # debug

    # 2. Parser (FIXED)
    parser = Parser(tokens)
    parsed = parser.parse()

    print("PARSED OUTPUT:", parsed)  # debug
    print("TYPE:", type(parsed))  # debug

    # 🔥 SAFETY CHECK (IMPORTANT)
    if not isinstance(parsed, dict):
        raise Exception(f"Parser did not return valid equation: {parsed}")

    # 3. Solver
    solutions = solve(parsed)

    print("TOTAL SOLUTIONS:", len(solutions))

    for s in solutions[:10]:  # only first 10
        print(s)

    # 4. Constraints
    if constraints:
        solutions = apply_constraints(solutions, constraints)

    return solutions

File: backend/core/engine.cpp
#include "engine.h"
#include "constraints.h"
#include "parser.h"
#include "solver.h"

#include <iostream>

namespace quant {

EngineResult solve_equation(
    const std::string &expression,
    const std::vector<std::string> &constraints,
    int displayLimit) {
    Parser parser;
    ParseResult parseResult = parser.parseEquation(expression);

    std::vector<Constraint> parsedConstraints = parseConstraints(constraints);
    auto validator = [&parsedConstraints](const std::map<std::string, int> &solution) {
        return isValid(solution, parsedConstraints);
    };

    Solver solver;
    SolverResult solverResult = solver.solve(parseResult.coefficients, parseResult.target, displayLimit, validator);

    if (solverResult.totalSolutions == 0) {
        throw std::runtime_error("No whole number solutions exist.");
    }

    EngineResult output;
    output.totalSolutions = solverResult.totalSolutions;
    output.shownSolutions = static_cast<int>(solverResult.solutions.size());
    output.solutions = std::move(solverResult.solutions);
    return output;
}

} // namespace quant
